src/mrpro/algorithms/reconstruction/IterativeSenseReconstruction.py

```diff
@@ -0,0 +1,148 @@
+        n_iterations: int,
+        csm: CsmData | None = None,
+        noise: None | KNoise = None,
+        dcf: DcfData | None = None,
+    ) -> None:
+        """Initialize DirectReconstruction.
+
+        Parameters
+        ----------
+        fourier_op
+            Instance of the FourierOperator which adjoint is used for reconstruction.
+        csm
+            Sensitivity maps for coil combination
+        n_iterations
+            Number of CG iterations
+        noise
+            Used for prewhitening
+        dcf
+            Density compensation. If None, no dcf will be performed.
+            Also set to None, if the FourierOperator is already density compensated.
+        """
+        super().__init__()
+        self.fourier_op = fourier_op
+        self.n_iterations = n_iterations
+        # TODO: Make this buffers once DataBufferMixin is merged
+        self.csm = csm
+        self.noise = noise
+        self.dcf = dcf
+
+    @classmethod
+    def from_kdata(cls, kdata: KData, n_iterations: int, noise: KNoise | None = None) -> Self:
+        """Create a IterativeSenseReconstruction from kdata with default settings.
+
+        Parameters
+        ----------
+        kdata
+            KData to use for trajektory and header information
+        n_iterations
+            Number of CG iterations
+        noise
+            KNoise used for prewhitening. If None, no prewhitening is performed
+        """
+        if noise is not None:
+            kdata = prewhiten_kspace(kdata, noise)
+        dcf = DcfData.from_traj_voronoi(kdata.traj)
+        fourier_op = FourierOp.from_kdata(kdata)
+        recon = DirectReconstruction(fourier_op, dcf=dcf, noise=noise)
+        image = recon.direct_reconstruction(kdata)
+        csm = CsmData.from_idata_walsh(image)
+        return cls(fourier_op, n_iterations, csm, noise, dcf)
+
+    def forward(self, kdata: KData) -> IData:
+        """Apply the reconstruction.
+
+        Parameters
+        ----------
+        kdata
+            k-space data to reconstruct.
+
+        Returns
+        -------
+            the reconstruced image.
+        """
+        device = kdata.data.device
+        if self.noise is not None:
+            kdata = prewhiten_kspace(kdata, self.noise.to(device))
+
+        # Create A^H W
+        operator = self.fourier_op.H
+        if self.dcf is not None:
+            operator = operator @ self.dcf.as_operator()
+        if self.csm is not None:
+            operator = self.csm.as_operator().H @ operator
+
+        # Calculate b = A^H W y
+        (right_hand_side,) = operator.to(device)(kdata.data)
+
+        # Create H = A^H W A
+        operator = operator @ self.fourier_op
+        if self.csm is not None:
+            operator = operator @ self.csm.as_operator()
+        operator = operator.to(device)
+
+        img_tensor = cg(
+            operator, right_hand_side, initial_value=right_hand_side, max_iterations=self.n_iterations, tolerance=0.0
```

Comment:
do we prefer the fixed number of iterations here or would it be better touse a tolerance?

Reply:
Commonly we want to do early stopping with iterative SENSE. IMO we want to run the algorithm for a set number of iterations rather than up to a certain tolerance.